design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost API v1beta1
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes making certain backward incompatible changes to the
+BareMetalHost API and publishing the result as version `v1beta1`.
+
+## Motivation
+
+The BareMetalHost API is currently stuck at `v1alpha1`. Meanwhile, several
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- BareMetalHost API version `metal3.io/v1beta1`.
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API.
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, it will be populated after inspection.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+The new architecture handling will make it possible for a
+`PreprovisioningImage` to have an empty architecture requested (currently,
+it defaults to `x86_64`). We expect implementations to use its default
```

Comment:
Isn't this equivalent to BMO choosing a better default?

Reply:
BMO does not know the default.

Reply:
Why is it harder to teach BMO the default (and have an explicit API) than to teach the preprovisioning image controller the default?
The code for both of those exists in the same repo.

Reply:
At the very least, it will remove the freedom of a preprovisioning image controller to detect the architecture (e.g. the assisted one in OCP can use their InfraEnv resources). In the built-in case, it is the preprovisioning image controller that now has access to DEPLOY_KERNEL/RAMDISK.